### PR TITLE
Implement better logging as gin middleware

### PR DESCRIPTION
**Pull request by Raffo**
This PR implements an opinionated gin middleware for logging of http requests. It uses logrus and structured logging to achieve that. Fields are according to what was discussed in previous discussion about logging at Zalando. Feel free to suggest changes. 

Files in `example` folder can safely be ignored as they are generated but can give an idea of the end result. 

EDIT: needs squash.

### CODE REVIEW

File: middleware/logger.go [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+		c.Next()
+
+		end := time.Now()
+		latency := end.Sub(start)
+		clientIP := c.ClientIP()
+		method := c.Request.Method
+		statusCode := c.Writer.Status()
+		comment := c.Errors.ByType(gin.ErrorTypePrivate).String()
+
+		log.SetFormatter(&log.TextFormatter{DisableColors: true})
```
[the remark this answers is not in the document]
Comment (Raffo) [reply middleware/logger.go]: I know and I'm happy to change it if you want to, but it was done on purpose. I hate that logs look different in different cases and I really think that a server should just log in one way. Your call though, I sneaked this in to let you/us think. 

The thing that was set by ever request is clearly a mistake.

File: middleware/logger.go [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+		log.WithFields(log.Fields{
+			"status_code":   statusCode,
+			"response_time": latency,
```
[the remark this answers is not in the document]
Comment (Raffo) [reply middleware/logger.go]: +1 for the names. 

For `duration`, It's easier to query but a bit harder to read. I would build for readability and not for easiness of query in Scalyr given that this project is also not aware of any log parsing in Scalyr and might be useful in different format.

File: middleware/logger.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (mikkeloscar) [reply middleware/logger.go]: Yeah I also don't care for colors. I would be fine with disabling it as part of initializing the Service: https://github.com/mikkeloscar/gin-swagger/blob/master/templates/api.gotmpl#L129-L161

**Comment by Raffo** [comment]
I updated the PR to answer the comment. 
I have two additional points: 

1. Current output looks like that: 

```
mir:example raffo$ go build . && ./example --insecure-http
time="2017-09-19T15:27:34+02:00" level=info msg="Serving 'Cluster Registry - 0.0.1' on address :8080"
time="2017-09-19T15:27:37+02:00" level=info client="::1" duration=66ns method=GET path=/ status=404
```

Do we want quotes around all the values? It probably makes sense from my point of view and I'm not sure why this is not the case, I can have a look at this. 


2. To be able to re-generate the templates, which is the problem I was having in #3 , I had to `rm bindata.go` and then regenerate with `make generate` which I think it is called with `make` anyways but was unable to detect the changes of templates. What do you thing about adding `rm bindata.go` to `make clean` for example? I hadn't look closely on why it doesn't work to be honest, that's why I didn't add to the PR in the first place.

File: middleware/logger.go [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+import (
+	"time"
+
+	"github.com/gin-gonic/gin"
+	log "github.com/sirupsen/logrus"
+)
+
+// LogrusLogger instance a Logger middleware that uses logrus for logging.
+// The output is in structured logging with the following format:
+// time="2017-09-16T05:32:16+02:00" level=info client:="::1" method=GET path=/ response_time=78ns status_code=404"
```
Comment (mikkeloscar) [inline middleware/logger.go]: The example is now outdated :)

**Comment by mikkeloscar** [comment]
> Do we want quotes around all the values? It probably makes sense from my point of view and I'm not sure why this is not the case, I can have a look at this.

I think it's fine to only have quotes for values with special chars.

> To be able to re-generate the templates, which is the problem I was having in #3 , I had to rm bindata.go and then regenerate with make generate which I think it is called with make anyways but was unable to detect the changes of templates. What do you thing about adding rm bindata.go to make clean for example? I hadn't look closely on why it doesn't work to be honest, that's why I didn't add to the PR in the first place.

I see, appears to be a bug in the Makefile where it doesn't consider changed to the templates folder. Should just be fixed in a separate PR IMO.

**Comment by mikkeloscar** [comment]
:+1:

**Comment by Raffo** [comment]
👍

**Comment by ideahitme** [comment]
@Raffo vacation is for resting :P

**Comment by Raffo** [comment]
@ideahitme 

Well: 
- This is not a Google doc
- I didn't code anything for the last ~3 weeks
- Only slightly related to work (not sure how you came here)
- Doesn't impact team work nor does affect my time (coded at the airport at 5 am, reviewed after a fat lunch). 

Conclusion: 10/10 would do it again.

**Pull request merged**